### PR TITLE
Added KNOWN_ERRORS functionality for BGS::ShareError

### PR DESCRIPTION
**Pull request by jcq**
includes new PowerOfAttorneyFolderDenied error class;

### CODE REVIEW

File: lib/bgs/base.rb [hunk 1]
```diff
@@ -7,6 +7,7 @@
 require "savon"
 require "nokogiri"
 require "httpclient"
+require "pry"
```
Comment (pkarman) [inline lib/bgs/base.rb]: why this?

File: lib/bgs/errors.rb [hunk 2]
```diff
@@ -103,14 +103,23 @@ def initialize(message, code = nil)
     def ignorable?
       TRANSIENT_ERRORS.any? { |transient_error| message.include?(transient_error) }
     end
+
+    def self.error_type(message, code)
```
Comment (pkarman) [inline lib/bgs/errors.rb]: for class methods we are trying to use the
```ruby
class << self
  def foo
  end
end
```

syntax, at the top of the class definition.

Comment (pkarman) [reply lib/bgs/errors.rb]: the name of this method doesn't communicate much to me. Maybe `new_from_message` would be clearer?

File: lib/bgs/errors.rb [hunk 3]
```diff
@@ -103,14 +103,23 @@ def initialize(message, code = nil)
...
+      new_error = nil
+      KNOWN_ERRORS.each do |msg_str, error_class|
```
Comment (pkarman) [inline lib/bgs/errors.rb]: where is KNOWN_ERRORS defined?

ah I see it later. We should define constants at the top of the file. See https://rubystyle.guide/#classes-modules

File: spec/bgs/base_spec.rb [hunk 4]
```diff
@@ -141,7 +113,40 @@
...
+        expect(error.class).to eq BGS::PowerOfAttorneyFolderDenied
+        expect(error.message).to eq message
+        expect(error.code).to eq 500
+        expect(error).to_not be_ignorable
+      end
+    end
+  end
+end
+
+def default_soap_body(message)
```
Comment (pkarman) [inline spec/bgs/base_spec.rb]: I like this refactor. Maybe move the `def` block to the top of the file?

File: spec/bgs/base_spec.rb [hunk 5]
```diff
@@ -141,7 +113,40 @@
...
   end
+
+  context "when Savon::SoapFault with a known error" do
+    let(:message) { "Power of Attorney of Folder is none. Access to this record is denied." }
+    let(:response_body) { default_soap_body(message) }
+
+    it "raises a BGS::PowerOfAttorneyFolderDenied error" do
+      allow_any_instance_of(Savon::Client).to receive(:call).and_raise(soap_fault)
+
+      expect { bgs_base.test_request(:method) }.to raise_error do |error|
```
Comment (pkarman) [inline spec/bgs/base_spec.rb]: I like this test!

File: lib/bgs/base.rb (hunk 1 above)
Comment (jcq) [reply lib/bgs/base.rb]: Oops, forgot to remove that after some quick debugging

File: lib/bgs/errors.rb (hunk 2 above)
Comment (jcq) [reply lib/bgs/errors.rb]: Makes sense — updated both of those

File: lib/bgs/errors.rb (hunk 3 above)
Comment (jcq) [reply lib/bgs/errors.rb]: I agree, it's a bit of a weird placement; the reason for that was that if it's at the top, that comes before the definition of the class that it references (`BGS::PowerOfAttorneyFolderDenied`). I think that's why the VBMS repo references the classes via strings in their KNOWN_ISSUES object (using the `constantize` method to then reference the class). However, for whatever reason, `constantize` wasn't available in the BGS repo environment. I assume there's some difference in dependencies, but wasn't familiar enough with stuff to figure it out. Searches online say that it should just be there in a Rails environment, but it wasn't working for me.

Happy to adjust however you think best, however!

Comment (pkarman) [reply lib/bgs/errors.rb]: https://coderwall.com/p/lf5bfg/ruby-convert-string-to-class suggests this should work:

```ruby
Kernel.const_get('ExceptionClassName')
```

Comment (jcq) [reply lib/bgs/errors.rb]: Yup, that seems to have worked. I moved the `KNOWN_ISSUES` def inside the `SharedError` class next to the existing `TRANSIENT_ISSUES` definition. Let me know if you have any other updates!

File: spec/bgs/base_spec.rb [hunk 6]
```diff
@@ -83,24 +85,10 @@
       end
      end
   end
-
+
```
Comment (pkarman) [inline spec/bgs/base_spec.rb]: maybe some extra whitespace here? we don't have codeclimate on this repo to warn us.

**Pull request merged**